src/core_plugins/vega/public/data_model/vega_parser.js

```diff
@@ -0,0 +1,390 @@
+import _ from 'lodash';
+import * as vega from 'vega';
+import * as vegaLite from 'vega-lite';
+import schemaParser from 'vega-schema-url-parser';
+import versionCompare from 'compare-versions';
+import { EsQueryParser } from './es_query_parser';
+import hjson from 'hjson';
+import { Utils } from './utils';
+import { EmsFileParser } from './ems_file_parser';
+import { UrlParser } from './url_parser';
+
+const DEFAULT_SCHEMA = 'https://vega.github.io/schema/vega/v3.0.json';
+
+const locToDirMap = {
+  left: 'row-reverse',
+  right: 'row',
+  top: 'column-reverse',
+  bottom: 'column'
+};
+
+// If there is no "%type%" parameter, use this parser
+const DEFAULT_PARSER = 'elasticsearch';
+
+export class VegaParser {
+
+  constructor(spec, searchCache, timeCache, dashboardContext, serviceSettings) {
+    this.spec = spec;
+    this.hideWarnings = false;
+    this.error = undefined;
+    this.warnings = [];
+
+    const onWarn = this._onWarning.bind(this);
+    this._urlParsers = {
+      elasticsearch: new EsQueryParser(timeCache, searchCache, dashboardContext, onWarn),
+      emsfile: new EmsFileParser(serviceSettings),
+      url: new UrlParser(onWarn),
+    };
+  }
+
+  async parseAsync() {
+    try {
+      await this._parseAsync();
+    } catch (err) {
+      // if we reject current promise, it will use the standard Kibana error handling
+      this.error = Utils.formatErrorToStr(err);
+    }
+    return this;
+  }
+
+  async _parseAsync() {
+    if (this.isVegaLite !== undefined) throw new Error();
+
+    if (typeof this.spec === 'string') {
+      this.spec = hjson.parse(this.spec, { legacyRoot: false });
+    }
+    if (!_.isPlainObject(this.spec)) {
+      throw new Error('Invalid Vega spec');
+    }
+    this.isVegaLite = this._parseSchema();
+    this.useHover = !this.isVegaLite;
+
+    this._config = this._parseConfig();
+    this.hideWarnings = !!this._config.hideWarnings;
+    this.useMap = this._config.type === 'map';
+    this.renderer = this._config.renderer === 'svg' ? 'svg' : 'canvas';
+
+    this._setDefaultColors();
+    this._parseControlPlacement(this._config);
+    if (this.useMap) {
```

Comment:
It seems like we are anyway throwing an error later if we try to draw a map. Why do we do all those parsing here, instead of throwing here directly?

Reply:
This code is kept in sync with the plugin - so while yes, at this point map config is being validated but later errorer out, this simplifies future migration to the full map support inside the core. If I remove all the map parsing here, it might be a bit harder to keep the updates in sync to both. I could introduce some sort of a plugable parser thing too, but it seems it would overcomplicate parsing code even more.  Do you think this is a must have?